### ISSUE TITLE
Poor Performance of Spring Doc Both Standalone and When Integrated with Spring Data REST, Leading to Excessive API Documentation Generation Time

### ISSUE THREAD

**Issue by 5GSBin** [springdoc/springdoc-openapi#2365]
**Issue Description**
Our team has encountered severe performance issues while using SpringDoc in conjunction with Spring Data REST. Generating the API documentation takes an exceedingly long time, approximately 20 minutes, creating an illusion that the documentation cannot be generated and is forever loading. Additionally, the first-time generation of documentation by SpringDoc is particularly slow, taking about a minute.

**Steps to Reproduce**
- Integrate SpringDoc and Spring Data REST in a Spring Boot application.
- Define multiple models and repositories using Spring Data REST.
- Attempt to generate OpenAPI documentation.

**Probable Cause**
Initial analysis suggests that the performance bottleneck likely occurs when SpringDoc integrates with Spring Data REST, requiring the parsing of a large number of entities and relationships. This process is single-threaded and synchronous.

**Suggested Solutions**
1. **Lazy Loading**: Only generate a specific part of the API data when the user actually needs to view that portion of the documentation.
2. **Multi-threading**: Considering that the task of generating API documentation can be broken down into smaller tasks (e.g., each controller or entity class is a small task), could multi-threading be used to handle these tasks in parallel?
3. **Offline Generation**: Provide an option to allow developers to generate static JSON files for the API documentation during the project build phase. This way, dynamic generation would not be required during application runtime.

Please consider these suggestions to optimize the performance of SpringDoc when integrated with Spring Data REST. Thank you!

**Comment by uc4w6c** [comment]
Please provide a [Minimal, Reproducible Example - with HelloController](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces the issue.

This ticket will be closed and can be reopened if the relevant information are provided.

I don't know if this problem can be solved, but how about setting `springdoc.pre-loading-enabled: true`?

**Comment by shitikanth** [comment]
I also encountered this issue in a project with 200+ controllers - springdoc was taking over a minute to generate the API docs on first load.

From what I have analyzed, I think the reason springdoc is slow on large projects is that it calls `ModelsConverters::resolveAsResolvedSchema` (from `SpringDocAnnotationsUtils::extractSchema`) for every annotated type (request param, response body etc) for every controller method.

The key issue is that each call creates a new ModelConverterContextImpl. As a result, schemas for the same type are recomputed repeatedly using reflection. In my project, some classes were analyzed hundreds of times.

I am not particularly familiar with springdoc's codebase, but if were possible to create a single `ModelConverterContextImpl` instance and reuse it across invocations, that should result in a significant performance improvement.
